### PR TITLE
Label implements IEquatable<Label>

### PR DESCRIPTION
**Pull request by Foxtrek64**
Adds an implementation of `IEquatable<Label>` to `System.Reflection.Emit.Label`.

Fixes dotnet/corefx#32959
Relies on [dotnet/corecrl/pull/22938](https://github.com/dotnet/coreclr/pull/22938)

### CODE REVIEW

**Comment by danmosemsft** [comment]
Thanks @Foxtrek64 . Can you please add a simple test, that fails before your change and passes after? Something that calls `Equals` through `IEquatable<Label>`

**Comment by danmosemsft** [comment]
The tests are apparently in src\System.Reflection.Emit.ILGeneration\tests\Label\LabelEquals.cs

**Comment by Foxtrek64** [comment]
I was unable to run the unit tests through VS (kept throwing `System.BadImageFormatException: Bad IL Format`) even after a successful build of both CoreCLR and CoreFx.

So I relied on msbuild to do it for me. It seems happy enough.
```
PS m:\corefx\src\System.Reflection.Emit.ILGeneration\tests> dotnet msbuild /t:BuildAndTest
Microsoft (R) Build Engine version 16.0.385-preview+g966cdf2ac6 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  System.Reflection.Emit.ILGeneration.Tests -> m:\corefx\artifacts\bin\System.Reflection.Emit.ILGeneration.Tests\netcore
app-Debug\System.Reflection.Emit.ILGeneration.Tests.dll
  ----- start 23:12:02.02 ===============  To repro directly: =====================================================
  pushd m:\corefx\artifacts\bin\tests\System.Reflection.Emit.ILGeneration.Tests\netcoreapp-Windows_NT-Debug-x64\
  m:\corefx\artifacts\bin\testhost\netcoreapp-Windows_NT-Debug-x64\dotnet.exe xunit.console.dll System.Reflection.Emit.I
LGeneration.Tests.dll -xml testResults.xml -nologo -notrait category=nonnetcoreapptests -notrait category=nonwindowstest
s -notrait category=failing -notrait category=OuterLoop
  popd
ne)
    Discovered:  System.Reflection.Emit.ILGeneration.Tests (found 128 test cases)
    Starting:    System.Reflection.Emit.ILGeneration.Tests (parallel test collections = on, max threads = 8)
  0
  emitWriteLine
  False
    Finished:    System.Reflection.Emit.ILGeneration.Tests
  === TEST EXECUTION SUMMARY ===
     System.Reflection.Emit.ILGeneration.Tests  Total: 222, Errors: 0, Failed: 0, Skipped: 0, Time: 18.655s
  ----- end 23:14:29.01 ----- exit code 0 ----------------------------------------------------------
```

**Comment by karelz** [comment]
@GrabYourPitchforks @steveharter can you please review it?
cc @joshfree

**Comment by steveharter** [comment]
@Foxtrek64 can you rebase so CI can see the changes made to the CLR? Thanks.

```
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19128.2\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Reflection.Emit.Label' does not implement interface 'System.IEquatable<System.Reflection.Emit.Label>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Reflection.Emit.ILGeneration\src\System.Reflection.Emit.ILGeneration.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19128.2\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Reflection.Emit.ILGeneration\uap10.0.16299-Debug\System.Reflection.Emit.ILGeneration.dll' [D:\a\1\s\src\System.Reflection.Emit.ILGeneration\src\System.Reflection.Emit.ILGeneration.csproj]
```

**Comment by Foxtrek64** [comment]
I've done a couple rebases (the second one just to make sure I did it right) and the packing test is still failing. If you guys have any ideas on something I may have missed, I'm all ears.

**Comment by steveharter** [comment]
I see now the packaging error was just for `uap`.

@joperezr @ericstj do we need to add to an ApiCompatBaseline.uapaot.txt file to remove the packaging error?

**Comment by karelz** [comment]
@ericstj @safern can you please help with the question above?

**Comment by ericstj** [comment]
```
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19170.2\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Reflection.Emit.Label' does not implement interface 'System.IEquatable<System.Reflection.Emit.Label>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Reflection.Emit.ILGeneration\src\System.Reflection.Emit.ILGeneration.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19170.2\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Reflection.Emit.ILGeneration\uap10.0.16299-Debug\System.Reflection.Emit.ILGeneration.dll' [D:\a\1\s\src\System.Reflection.Emit.ILGeneration\src\System.Reflection.Emit.ILGeneration.csproj]
```

Note the `uap10.0.16299-Debug` in this path.  You've modified the reference assembly which applies to the last version of UAP (and older versions of .NETCoreApp...).  I think you need to put the change in https://github.com/dotnet/corefx/blob/d83698810f608688f103951a1d37ad749c8a3a5f/src/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.netcore.cs instead.

**Comment by Foxtrek64** [comment]
That file doesn't have a Label declaration. Should I add one, perhaps cloning the code from this declaration? https://github.com/dotnet/corefx/blob/d83698810f608688f103951a1d37ad749c8a3a5f/src/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs#L59-L67

**Comment by ericstj** [comment]
It’s an additional file, you cannot redefine members defined in another file.  Just make sure the class is partial and add the interface implementation in the additional file.  It’s just a handy way of doing ifdef’s if they can be represented using partial classes.

**Comment by Foxtrek64** [comment]
@ericstj Eric, could you clarify your last post? I'm not entirely certain what you mean.
"Just make sure the class is partial...." What class?
"...in the additional file." What additional file?

**Comment by ericstj** [comment]
>  What class?

Apologies, in this case its a struct: `Label`.  It is partial already: https://github.com/dotnet/corefx/blob/aa3865a16e4b53a58414a99dc9fcaaee653f59e0/src/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs#L59

> What additional file?

The file I suggested that you change.  https://github.com/dotnet/corefx/blob/d83698810f608688f103951a1d37ad749c8a3a5f/src/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.netcore.cs

In other words, don't modify System.Reflection.Emit.ILGeneration.cs, literally only add the following to System.Reflection.Emit.ILGeneration.netcore.cs:
```c#
public partial struct Label : IEquatable<System.Reflection.Emit.Label>
{ }
```

File: src/System.Reflection.Emit.ILGeneration/tests/Label/LabelEquals.cs [hunk 1]
```diff
@@ -49,5 +49,14 @@ public void Equals_EqualityOperators()
...
         }
+
+        [Fact]
+        public void Equals_IEquatable()
+        {
+            IEquatable<Label> label1 = new Label();
+            Label label2 = new Label();
+
+            Assert.True(label1.Equals(label2));
+        }
```
Comment (stephentoub) [inline src/System.Reflection.Emit.ILGeneration/tests/Label/LabelEquals.cs]: Nit: this test would pass if the implementation of `IEquatable<Label>.Equals` were just `=> true`.  Could we fill it out a bit more?  We could also just add relevant lines to existing test cases above instead of adding this test, e.g. add:
```C#
Assert.True(((IEquatable<Label>)label).Equals((IEquatable<Label>)label));
```
to the `Equals_SameInstance_ReturnsTrue` test, and adding:
```C#
Assert.False(((IEquatable<Label>)label1).Equals((IEquatable<Label>)label2));
```
to the `Equals_DifferentLabel_ReturnsFalse` test.

Comment (Foxtrek64) [reply src/System.Reflection.Emit.ILGeneration/tests/Label/LabelEquals.cs]: That's a fair point. I'll do some tinkering, though I do like your idea of merging with the already existing equality tests.

Comment (Foxtrek64) [reply src/System.Reflection.Emit.ILGeneration/tests/Label/LabelEquals.cs]: I've made the recommended changes, though I've removed the redundant casts.

**Comment by stephentoub** [comment]
Thanks.
